File: Dynamic_Programming/Money_Sums/Money_Sums.py
# Problem Name: Money_Sums
# Language: Python
# Category: Dynamic_Programming
# Date: 2026-07-12

import sys
 
data = sys.stdin.read().split()
 
n = int(data[0])
ar = [int(x) for x in data[1:]]
t = sum(ar)
dp = [False]*(t+1)
dp[0] = True
for a in ar :
    for i in range(t,a-1,-1):
        if dp[i-a]:
            dp[i] = True
o = [i for i in range(i,t+1) if dp[i] ]
print(len(o))
print(*o)

File: Dynamic_Programming/Money_Sums/Money_Sums.cpp
/*
 * Problem Name: Money_Sums
 * Language: C++
 * Category: Dynamic_Programming
 * Date: 2026-07-12
 */

#include<bits/stdc++.h>
using namespace std;
 
int main(){
    ios_base::sync_with_stdio(0);
    cin.tie(nullptr);
 
    int n ;
    cin>>n;
    cin>>ws;
    vector<int> arr(n+1,0);
    int total = 0;
    for(int i = 1 ; i <= n ; i++){
        cin>>arr[i];
        total += arr[i];
    }
 
    vector<vector<bool>> dp(n+1,vector<bool>(total+1,0));
    dp[0][0] = 1;
 
    for(int i = 1 ; i <= n ; i++){
        for(int j = 0 ; j <= total ; j++){
            dp[i][j] = dp[i-1][j];
            if (j >= arr[i]){
                dp[i][j]  = dp[i][j] || dp[i-1][j-arr[i]];  
            }
        }
    }
 
    int ans = 0 ;
    for(int i = 1 ; i <= total ; i++){
        if (dp[n][i]) ans++;
    }
    cout<<ans<<"\n";
    for(int i = 1 ; i <= total ; i++){
        if (dp[n][i]) cout<<i<<" ";
    }
    cout<<endl;
    return 0;
}